### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by jetuk**
Two performance improvements related to pytables.

- First is an easy win the `TablesRecorder`
- Second is cythoning the `TablesArrayParameter` this also reads the entire array on init. It's not ideal for indices due to have to type to `double` and cast back to `int`.

### CODE REVIEW

**Comment by snorfalorpagus** [comment]
Is there anything we can do to avoid the casting? Check the type of the node, hand have a `_values_double` and `_values_int` memoryview? Or does this complicate `value()` too much?

**Comment by jetuk** [comment]
Use two classes.

**Comment by snorfalorpagus** [comment]
> Use two classes.

Instead we could have `value()` read the doubles array and `index()` read the integer array. If the respective array is `None` then cast the value from the other array.

File: pywr/recorders.py [hunk 1]
```diff
@@ -248,23 +248,29 @@ def save(self):
...
             if isinstance(node, AbstractStorage):
                 ca[idx, :] = node.volume
             elif isinstance(node, AbstractNode):
                 ca[idx, :] = node.flow
             elif isinstance(node, IndexParameter):
+                a = np.empty(len(self.model.scenarios.combinations), dtype=np.int)
                 for si in self.model.scenarios.combinations:
-                    ca[idx, si.global_id] = node.index(ts, si)
+                     a[si.global_id] = node.index(ts, si)
+                ca[idx, :] = a
```
Comment (jetuk) [inline pywr/recorders.py]: Each call to `ca[idx, si.global_id]` creates lots of overhead in pytables. Doing it once saves this.

**Comment by jetuk** [comment]
That sounds a much better solution.

**Comment by snorfalorpagus** [comment]
I've made the change. There are existing tests that use both the `value()` and `index()` method of `TablesArrayParameter`, so I think this is good to go (once you review my changes).

File: pywr/parameters/_parameters.pyx [hunk 2]
```diff
@@ -261,6 +262,111 @@ cdef class ArrayIndexedScenarioParameter(Parameter):
...
+    cpdef reset(self):
+        self.h5store = H5Store(self.h5file, None, "r")
+        node = self.h5store.file.get_node(self.where, self.node)
+
+        # detect data type and read into memoryview
+        if node.dtype == np.float32 or node.dtype == np.float64:
+            self._values_dbl = node.read()
+            self._values_int = None
+            shape = self._values_dbl.shape
+        else:
```
Comment (jetuk) [inline pywr/parameters/_parameters.pyx]: Should we be explicit about which types to use as ints here? And throw an error it's something else?

Comment (snorfalorpagus) [inline pywr/parameters/_parameters.pyx]: Yes. I'll make the change.

**Pull request merged**